// @progress/kendo-dropdowntree-vue-wrapper/dist/es/KendoDropDownTree/KendoDropDownTreeMixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _kendoBaseComponentsVueWrapper = require('@progress/kendo-base-components-vue-wrapper');

var _DropDownTreeProps = require('./DropDownTreeProps');

var _DropDownTreeProps2 = _interopRequireDefault(_DropDownTreeProps);

var _Animation = require('./Animation');

var _Animation2 = _interopRequireDefault(_Animation);

var _Checkboxes = require('./Checkboxes');

var _Checkboxes2 = _interopRequireDefault(_Checkboxes);

var _Messages = require('./Messages');

var _Messages2 = _interopRequireDefault(_Messages);

var _Popup = require('./Popup');

var _Popup2 = _interopRequireDefault(_Popup);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    props: _DropDownTreeProps2.default,
    render: function render(createElement) {
        var tagName = this.$props.checkboxes || this.$props.checkboxesCheckChildren || this.$props.checkboxesName || this.$props.checkboxesTemplate ? 'select' : 'input';

        return createElement(tagName, [this.$slots['kendo-hierarchicaldatasource']]);
    },

    model: {
        event: 'changemodel'
    },
    mixins: [_kendoBaseComponentsVueWrapper.KendoBaseComponent, _kendoBaseComponentsVueWrapper.KendoBaseDatasourceComponent, _Animation2.default, _Checkboxes2.default, _Messages2.default, _Popup2.default],
    mounted: function mounted() {
        var that = this;

        kendo.jQuery(this.$el).kendoDropDownTree(that.widgetOptions);

        that.$emit('kendowidgetready', that.kendoWidget());

        that.kendoWidget().bind('change', function () {
            that.$emit('changemodel', that.kendoWidget().value());
        });
    },

    methods: {
        kendoWidget: function kendoWidget() {
            return kendo.jQuery(this.$el).getKendoDropDownTree();
        }
    }
};
