// @progress/kendo-dropdowntree-vue-wrapper/dist/npm/KendoDropDownTree/DropDownTreeProps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = {
    animation: Boolean | Object,
    autoBind: Boolean,
    autoClose: Boolean,
    autoWidth: Boolean,
    checkAll: Boolean,
    checkAllTemplate: String | Function,
    checkboxes: Boolean | Object,
    clearButton: Boolean,
    dataImageUrlField: String,
    dataSource: Object | Array,
    dataSpriteCssClassField: String,
    dataTextField: String | Array,
    dataUrlField: String,
    dataValueField: String | Array,
    delay: Number,
    enable: Boolean,
    enforceMinLength: Boolean,
    filter: String,
    footerTemplate: String | Function,
    height: String | Number,
    ignoreCase: Boolean,
    loadOnDemand: Boolean,
    messages: Object,
    minLength: Number,
    noDataTemplate: String | Function,
    placeholder: String,
    popup: Object,
    headerTemplate: String | Function,
    valueTemplate: String | Function,
    tagMode: String,
    template: String | Function,
    text: String,
    value: String | Array,
    valuePrimitive: Boolean,

    // Events
    change: Function,
    close: Function,
    dataBound: Function,
    filtering: Function,
    open: Function,
    select: Function
};
